plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs

```diff
@@ -422,6 +431,43 @@ hoistExpr var t =
                     (Set.map LexName deps)
                 pure $ PIR.mkVar () var'
 
+mkTrace
+    :: (PLC.Contains uni T.Text)
+    => PLC.Type PLC.TyName uni ()
+    -> T.Text
+    -> PIRTerm uni PLC.DefaultFun
+    -> PIRTerm uni PLC.DefaultFun
+mkTrace ty str v =
+    PLC.mkIterApp
+        ()
+        (PIR.TyInst () (PIR.Builtin () PLC.Trace) ty)
+        [PLC.mkConstant () str, v]
+
+traceInside ::
+    T.Text
+    -> PIR.Name
+    -> PIRTerm PLC.DefaultUni PLC.DefaultFun
+    -> PLC.Type PIR.TyName PLC.DefaultUni ()
+    -> PIRTerm PLC.DefaultUni PLC.DefaultFun
+traceInside varName lamName (LamAbs () n t body) (PLC.TyFun () _dom cod) =
+    -- when t = \x -> body, => \x -> traceInside body
+    LamAbs () n t (traceInside varName lamName body cod)
```

Comment:
nit: since `varName` and `lamName` are constant until later, I'd probably write this with a helper function. Something like:
```
traceInside varName lamName = go
  where 
    go (LamAbs ...) (PLC.TyFun ...) = ...
```